File: src/calculators.js
function parseNumber(value) {
  const number = Number(value);
  return Number.isFinite(number) ? number : 0;
}

function calcRD(monthly, annualRate, months) {
  const M = parseNumber(monthly);
  const rA = parseNumber(annualRate);
  const N = parseNumber(months);
  const r = rA / 400;
  let maturity = 0;

  if (M > 0 && N > 0) {
    for (let i = 1; i <= N; i += 1) {
      maturity += M * Math.pow(1 + r, (N - i + 1) / 3);
    }
  }

  const invested = M * N;
  const interest = Math.max(0, maturity - invested);

  return { maturity, invested, interest };
}

function calcFD(principal, annualRate, years) {
  const P = parseNumber(principal);
  const rA = parseNumber(annualRate);
  const Y = parseNumber(years);
  const quarterly = 4;
  const r = rA / 100;

  const maturity = P > 0 && Y > 0 ? P * Math.pow(1 + r / quarterly, quarterly * Y) : 0;
  const interest = Math.max(0, maturity - P);

  return { maturity, principal: P, interest };
}

function calcEMI(principal, annualRate, years) {
  const P = parseNumber(principal);
  const rA = parseNumber(annualRate);
  const Y = parseNumber(years);
  const monthlyRate = rA / 12 / 100;
  const n = Y * 12;
  let emi = 0;

  if (P > 0 && n > 0) {
    emi = monthlyRate > 0
      ? (P * monthlyRate * Math.pow(1 + monthlyRate, n)) / (Math.pow(1 + monthlyRate, n) - 1)
      : P / n;
  }

  const total = emi * n;
  const interest = Math.max(0, total - P);

  return { monthly: emi, principal: P, interest, total };
}

function calcNW(assets, liabilities) {
  const A = parseNumber(assets);
  const L = parseNumber(liabilities);
  const netWorth = Math.max(0, A - L) || A - L;

  return { netWorth, assets: A, liabilities: L };
}

function calcSIP(monthly, annualRate, years) {
  const M = parseNumber(monthly);
  const monthlyRate = parseNumber(annualRate) / 1200;
  const months = Math.floor(parseNumber(years) * 12);
  const invested = M * months;
  const maturity = M > 0 && months > 0
    ? monthlyRate > 0
      ? M * ((Math.pow(1 + monthlyRate, months) - 1) / monthlyRate) * (1 + monthlyRate)
      : invested
    : 0;
  return { maturity, invested, interest: Math.max(0, maturity - invested), months };
}

function calcAmortization(principal, annualRate, years) {
  const summary = calcEMI(principal, annualRate, years);
  const rate = parseNumber(annualRate) / 1200;
  const months = Math.floor(parseNumber(years) * 12);
  let balance = summary.principal;
  const schedule = [];
  for (let month = 1; month <= months; month += 1) {
    const interest = balance * rate;
    const principalPaid = Math.min(balance, Math.max(0, summary.monthly - interest));
    balance = Math.max(0, balance - principalPaid);
    schedule.push({ month, payment: summary.monthly, principal: principalPaid, interest, balance });
  }
  return { ...summary, schedule };
}

function calcLoanComparison(loanA, loanB) {
  const a = calcAmortization(loanA.principal, loanA.annualRate, loanA.years);
  const b = calcAmortization(loanB.principal, loanB.annualRate, loanB.years);
  return { loanA: a, loanB: b, betterOption: a.total <= b.total ? 'loanA' : 'loanB', savings: Math.abs(a.total - b.total) };
}

function calcInsurance(coverage, term, age, healthFactor) {
  const C = parseNumber(coverage);
  const A = parseNumber(age);
  const factor = Math.max(0.5, parseNumber(healthFactor) || 1);
  const premium = C * 0.02 * Math.max(1, (A - 25) / 10) * factor / 100;
  return { coverage: C, term: parseNumber(term), age: A, healthFactor: factor, annualPremium: premium, monthlyPremium: premium / 12 };
}

function calcBudget(income, expenses, savingsTarget) {
  const I = parseNumber(income);
  const normalizedExpenses = Object.fromEntries(Object.entries(expenses || {}).map(([key, value]) => [key, Math.max(0, parseNumber(value))]));
  const totalExpenses = Object.values(normalizedExpenses).reduce((sum, value) => sum + value, 0);
  const targetSavings = Math.max(0, Math.min(I, I * parseNumber(savingsTarget) / 100));
  return { income: I, expenses: normalizedExpenses, totalExpenses, targetSavings, surplus: I - totalExpenses - targetSavings };
}

function calcRetirement(currentAge, retirementAge, currentSavings, monthlyContribution, annualRate) {
  const years = Math.max(0, parseNumber(retirementAge) - parseNumber(currentAge));
  const months = Math.floor(years * 12);
  const rate = parseNumber(annualRate) / 1200;
  let futureValue = parseNumber(currentSavings);
  for (let month = 0; month < months; month += 1) futureValue = (futureValue + parseNumber(monthlyContribution)) * (1 + rate);
  const invested = parseNumber(currentSavings) + parseNumber(monthlyContribution) * months;
  return { years, futureValue, invested, growth: Math.max(0, futureValue - invested) };
}

module.exports = { calcRD, calcFD, calcEMI, calcNW, calcSIP, calcAmortization, calcLoanComparison, calcInsurance, calcBudget, calcRetirement };

File: py_src/calculators.py
from __future__ import annotations

from typing import Any


def parse_number(value: Any) -> float:
    try:
        number = float(value)
        return number if number == number and number != float('inf') else 0.0
    except (TypeError, ValueError):
        return 0.0


def calc_rd(monthly: Any, annual_rate: Any, months: Any) -> dict[str, float]:
    m = parse_number(monthly)
    r_a = parse_number(annual_rate)
    n = int(parse_number(months))
    r = r_a / 400.0
    maturity = 0.0

    if m > 0 and n > 0:
        for i in range(1, n + 1):
            maturity += m * ((1 + r) ** ((n - i + 1) / 3.0))

    invested = m * n
    interest = max(0.0, maturity - invested)
    return {
        "maturity": maturity,
        "invested": invested,
        "interest": interest
    }


def calc_fd(principal: Any, annual_rate: Any, years: Any) -> dict[str, float]:
    p = parse_number(principal)
    r_a = parse_number(annual_rate)
    y = parse_number(years)
    quarterly = 4.0
    r = r_a / 100.0

    maturity = p * ((1 + r / quarterly) ** (quarterly * y)) if p > 0 and y > 0 else 0.0
    interest = max(0.0, maturity - p)
    return {
        "maturity": maturity,
        "principal": p,
        "interest": interest
    }


def calc_emi(principal: Any, annual_rate: Any, years: Any) -> dict[str, float]:
    p = parse_number(principal)
    r_a = parse_number(annual_rate)
    y = parse_number(years)
    monthly_rate = r_a / 12.0 / 100.0
    n = y * 12.0
    emi = 0.0

    if p > 0 and n > 0:
        emi = (
            (p * monthly_rate * ((1 + monthly_rate) ** n)) /
            (((1 + monthly_rate) ** n) - 1)
        ) if monthly_rate > 0 else p / n

    total = emi * n
    interest = max(0.0, total - p)
    return {
        "monthly": emi,
        "principal": p,
        "interest": interest,
        "total": total
    }


def calc_nw(assets: Any, liabilities: Any) -> dict[str, float]:
    a = parse_number(assets)
    l = parse_number(liabilities)
    net_worth = a - l
    return {
        "netWorth": net_worth,
        "assets": a,
        "liabilities": l
    }
